Connection manager

dbrgn opened the pull request:
Early draft of a ConnectionManager that would address the reconnect behavior discussed in #218.

Would be great to get some feedback on it. @Marwes you've written something like this in the past, right?

To test, run `cargo run --example async-connection-loss --features="tokio-rt-core" reconnect` and then restart and/or stop redis. Output for a 2s server downtime:

```
PING
Query result: Ok("PONG")
PING
Query result: Ok("PONG")
PING
Connection lost, reconnecting
Reconnecting failed: Connection refused (os error 111)
Query result: Err(broken pipe)
PING
Connection lost, reconnecting
Reconnecting failed: Connection refused (os error 111)
Query result: Err(broken pipe)
PING
Connection lost, reconnecting
Query result: Err(broken pipe)
PING
Query result: Ok("PONG")
```

I'll add some comments inline.

Based on #276.

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+    pub struct ConnectionManager {
+        connection_info: ConnectionInfo,
+        connection: MultiplexedConnection,
+    }
```
The `ConnectionManager` wraps a `MultiplexedConnection`. I still have to add cloning behavior to match the `MultiplexedConnection`, but the idea was that the manager can be passed around to different tasks / connection handlers just like the `MultiplexedConnection` itself.

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+    impl ConnectionManager {
+        /// Connect to the server and store the connection inside the returned `ConnectionManager`.
+        ///
+        /// TODO: Add `new_tokio`?
```
I'm not sure if this could easily be implemented non-tokio-specific, the futures from the old and the new connection would need to be proxied somehow. If that's possible, we could split `new` and `new_tokio`.

Marwes replied:
Just need to propagate `driver` out to the caller again for this.

dbrgn replied:
But the `driver` is replaced when reconnecting.

Marwes replied:
Aha, that's true. Could perhaps use https://docs.rs/futures/0.3.1/futures/task/trait.Spawn.html but tokio unfortunately does not implement it.

dbrgn replied:
Ok, then I guess we have to require Tokio for now.

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+            let (new_connection, driver) = MultiplexedConnection::new(connect(&self.connection_info).await?);
+            tokio::spawn(driver);
+            let _old_connection = std::mem::replace(&mut self.connection, new_connection);
+            Ok(self)
```
This seems quite straightforward. The old connection will be dropped, resulting in cleanup. That should work, right?

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+                        println!("Connection lost, reconnecting");
+                        if let Err(reconnect_failed) = self.reconnect().await {
+                            println!("Reconnecting failed: {}", reconnect_failed);
+                        };
```
The current logic will trigger a reconnect attempt once after every "connection dropped" error.

It means that right now multiple reconnection attempts could happen in parallel if multiple errors happen simultaneously (which is very likely).

Does someone have an idea on how to solve that cleanly?

Marwes replied:
Store something that uniquely identifies each individual connection and then when calling `reconnect` you check if the connection is the same or has changed. If it has changed then a concurrent request has already started the reconnection and nothing needs to be done.

Could be an integer incremented on every reconnection attempt or you could store the connection behind an `Arc` and compare the pointers.

dbrgn replied:
Ah, @Marwes mentioned `future::Shared` here: https://docs.rs/futures/0.3.1/futures/future/trait.FutureExt.html#method.shared That could be useful.

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+                        // attempts being done in parallel... We should
+                        // probably avoid that.
+                        println!("Connection lost, reconnecting");
+                        if let Err(reconnect_failed) = self.reconnect().await {
```
This should probably happen asynchronously, but I had a lifetime issue when I put the `self.reconnect()` call into `tokio::spawn`.

Marwes replied:
You will need to put the contents of `ConnectionManager` behind an `Arc`. As it is you can only issue one request at a time on the `ConnectionManager`.

That will then let you clone `self` when needed to get a 'static value for `tokio::spawn`.

Marwes commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+        async fn reconnect(&mut self) -> RedisResult<&mut Self> {
+            let (new_connection, driver) = MultiplexedConnection::new(connect(&self.connection_info).await?);
+            tokio::spawn(driver);
+            let _old_connection = std::mem::replace(&mut self.connection, new_connection);
```
self.connection = new_connection;

dbrgn replied:
D'oh :sweat_smile:

Marwes commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+                        // attempts being done in parallel... We should
+                        // probably avoid that.
+                        println!("Connection lost, reconnecting");
+                        self.reconnect().await;
```
This waits for the new connection to successfully reconnect. It should instead start the reconnection (`spawn` it on its own task) and just let this request return the error it encountered. This is where https://docs.rs/futures/0.3.1/futures/future/trait.FutureExt.html#method.shared are useful.  Share the reconnection future, then clone and spawn it once while storing the `Shared` future in the "Manager" `connection_future: Shared<...>`. Then `self.connection_future.await?` can be used to retrieve the new connection by just waiting for it to reconnect.

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
+                        };
+                    }
+                }
+                result
```
The error will be passed on to the caller unmodified. I think this is the only feasible generic solution, since depending on the application logic, if the connection fails after redis has processed the request, automatic retrying might have unintended consequences.

This way, the application must deal with the error, but reconnecting will happen automatically.

dbrgn commented:
I have to say I'm a bit stuck here.

In the current version of the code, `ConnectionManager` is not yet cloneable. Once `#[derive(Clone)]` is enabled, it's possible to clone the `ConnectionManager`, which is fine in general because it wraps a `Shared` instance.

However, once a manager instance is cloned, if instance A replaces the shared future with the reconnect future, instance B will still have a reference to the old shared future, right?

That would mean that we would need some kind of wrapper type with interior mutability. For example an Arc:

```rust
#[derive(Clone)]
pub struct ConnectionManager(Arc<ConnectionManagerInner>);

struct ConnectionManagerInner {
    connection_info: ConnectionInfo,
    connection: MultiplexedConnection,
}
```

Now references to the connection need to be updated. For read-only references that's easy:

```rust
fn get_db(&self) -> i64 {
    self.0.connection.db
}
```

However, mutable references now complain that the arc contents cannot be borrowed mutably. For example, the reconnect method now cannot replace the connection:

```rust
error[E0594]: cannot assign to data in an `Arc`
   --> src/aio.rs:620:13
    |
620 |             self.0.connection = new_connection;
    |             ^^^^^^^^^^^^^^^^^ cannot assign
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `std::sync::Arc<aio::connection_manager::ConnectionManagerInner>`
```

I could now wrap the entire `ConnectionManagerInner` in something like a `Mutex` or a `RwLock`, but then we lose the advantages that we got from wrapping `MultiplexedConnection`.

Am I on the wrong path? :slightly_smiling_face:

Marwes commented:
No, you will need some sort of interior mutability. https://docs.rs/arc-swap/0.4.4/arc_swap/index.html works well and you can use `compare_and_swap` to prevent multiple reconnections.

dbrgn commented:
@Marwes thanks for pointing out `arc_swap`, looks great! However, `req_packed_command` takes `&mut self` as argument. This would mean that I'd either have to lock the `ArcSwap` in write mode for every message dispatched, or to clone the `MultiplexedConnection` for every message dispatched.

ArcSwap about their performance characteristics:

> Write operations are considered expensive. A write operation is more expensive than access to an uncontended Mutex and on some architectures even slower than uncontended RwLock.

Cloning the `MultiplexedConnection` would mean cloning an `mpsc::Sender` as well as the db `i64`. Does that sound reasonable? It would be great if we could opt-in to automatic reconnection without having to pay that cost, but I'm not sure if it's possible. (Caching the cloned connection would mean that the reference could become stale...)

Marwes commented:
Cloning a `MultiplexedConnection` is pretty cheap, it might be possible to do better though by adding methods which work with `&self` which this could call. But it would need https://github.com/tokio-rs/tokio/pull/2041/files in that case.

Marwes commented:
Perhaps it could cache the connection, but still do a read on the `arc_swap` to check that the connection isn't stale?

dbrgn commented on src/aio.rs:
```diff
@@ -577,3 +577,111 @@ impl ConnectionLike for MultiplexedConnection {
+            tokio::spawn(driver);
+
+            // Update the connection in the connection manager
+            self.connection.compare_and_swap(guard, Arc::new(new_connection));
```
@Marwes this is not sufficient to avoid creating multiple connections in parallel, right? The way I understand things, it will only prevent *storing* multiple connections (overwriting the previously established connection). In a multi-threaded scenario, the `.await` call on the `connect()` method may yield control to another task that also launches a connection.

If that is correct, I will add an atomic connection counter to avoid reconnecting multiple times.

I didn't manage to trigger multiple connections when removing the loop delay in the example, but that's probably just because I did not enqueue reconnection tasks from multiple threads at the same time.

Marwes replied:
We can't prevent creating the future twice, but it is possible to prevent the reconnection from happening twice. Something like this.

```rust
// We can reuse the guard/connection reference that were used to run the connection
fn reconnect(&self,  current: arc_swap::Guard<_>) {
    let (new_connection, driver) = MultiplexedConnection::new(
        connect(&self.connection_info).await?
    );

    // Update the connection in the connection manager
    let prev = self.connection.compare_and_swap(guard, Arc::new(new_connection));

    if Arc::ptr_eq(&prev, current) {
        tokio::spawn(driver);
        tokio::spawn(new_connection);
    }
}
```

dbrgn replied:
Hmm, but `connect(...).await` will already open the TCP socket, right?

dbrgn replied:
Also, `new_connection` is a pipeline. Does it need to be spawned? I thought only the driver needs to be spawned.

Marwes replied:
```
    let new_connection =  async move {
        let (new_connection, driver) = MultiplexedConnection::new(
            connect(&self.connection_info).await?
        );
        tokio::spawn(driver);
        Ok(new_connection)
    }.shared();

    // Update the connection in the connection manager
    let prev = self.connection.compare_and_swap(guard, Arc::new(new_connection));

    if Arc::ptr_eq(&prev, current) {
        // start the connection attempt immediately but do not wait on it
        tokio::spawn(new_connection);
    }
```

dbrgn replied:
Ok, that makes sense. However, now the `ArcSwap` doesn't wrap a `MultiplexedConnection` anymore, but a `dyn Future<MultiplexedConnection>` instead: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6ecee2780e2046452967f70fa97d5f3e

I tried putting a `Shared` in the `ArcSwap` instead. However, with static dispatch this does of course not work due to a type change in the future. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=36e7e78a91db04a403c06ec379c565e6

I tried to use dynamic dispatch instead but couldn't get it to work :confused: I'll try again tomorrow, but if you have any pointer that would be helpful. (These are the situations where I sometimes still bump into the Rust async APIs, even though I've done async stuff for a while now.)

Marwes replied:
You will indeed need to box it and use dynamic dispatch https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=99941c415672f38d43b81c7b7cb67b3b

dbrgn replied:
Great, thanks! Somehow it did not occur to me to put the pinned box *inside* the Shared instance.

I also managed to simplify the code a bit. `boxed()` already returns a pinned box, so the type annotations and `Box::new` calls are not necessary. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b62964c32866c8ab11af0912eff1d4c9

dbrgn replied:
Hmm, it seems that since reconnecting can fail, I'll have to wrap the `RedisError` in an `Arc` since it must be `Clone` to be put inside `Shared`...

dbrgn replied:
@Marwes I hit another blocker: When a reconnection starts, the future is stored. All commands that are issued after that point in time should await the connection future before they can issue the command.

The problem is that `Shared` cannot contain a `Future` that resolves to a `RedisError` because that is not `Clone`. I solved this by using `Result<T, Arc<RedisError>>` instead, which works generally but leads to the following issue: If `req_packed_command` fails, it must return a `RedisError`. From the reconnection we get an `Arc<RedisError>` but cannot clone the error (since it's not `Clone`).

I see three possibilities:

- Put `io::Error` inside `RedisError` into something like an `Arc` so that it can be cloned, or use a cloneable abstraction [like this one (seems unmaintained)](https://crates.io/crates/arc-io-error) instead. However, that would be a breaking change.
- Convert `ErrorRepr::IoError(io::Error)` into `ErrorRepr::WithDescription(ErrorKind::IoError, "Reconnecting failed")` and discard the error details. (They should probably be logged though, see #277.) Main downside of this is that errors returned by initial connection and reconnection are different (since error details are discarded). The user does not see *why* the reconnection failed (except in logs). Note that a failed reconnection could immediately trigger another attempt.
- Change the implementation logic. As long as the connection is not yet established, fail all commands immediately with `ErrorRepr::WithDescription(ErrorKind::IoError, "Connection lost, reconnection in progress") until the connection is re-established.

I tend towards number 2. What do you think?

dbrgn replied:
I implemented number 2 for now. Seems to work! We can also pass in a string representation of the I/O error to the cloned error.

Marwes replied:
I think ideally we would return the original `io::Error` to the first one to see the error, while the rest get a generic `Reconnecting failed` one. May need to use a `Mutex` for that though, unsure if there is a better way than that :/

dbrgn commented:
I managed to get things working with `arc_swap`. Very nice library, thanks for the recommendation! Also thanks for the help with the pull request :slightly_smiling_face: 

I'll work on the remaining open issues tomorrow:

- Ensure that multiple connections are not established in parallel
- Reconnect asynchronously (don't `.await` after reconnecting)

Open questions:

- Is triggering a reconnect attempt on every failed command OK, or should there be some kind of throttling? Throttling might result in more failed connection attempts than not doing it, but at the same time the server is not hammered with connection attempts.
- Is the separate `connection-manager` feature good? Should it be named differently?
- This is probably non-trivial to unit-test, right?

Marwes commented:
> This is probably non-trivial to unit-test, right?

You could abstract `ConnectionManager<T: ConnectionLike>`, and supply a mocked connection instead of `MultiplexedConnection`

Marwes commented:
> Is triggering a reconnect attempt on every failed command OK, or should there be some kind of throttling? Throttling might result in more failed connection attempts than not doing it, but at the same time the server is not hammered with connection attempts.

A circuitbreaker could be built on top of this but the issue comes with figuring out a good strategy for it or making the strategy pluggable which is complicated in its own right :/

> Is the separate connection-manager feature good? Should it be named differently?

Sounds good to me.

dbrgn commented:
> A circuitbreaker could be built on top of this but the issue comes with figuring out a good strategy for it or making the strategy pluggable which is complicated in its own right :/

Sounds like something that should ideally be implemented in a separate library, so that other projects could re-use it. Probably out of scope for this PR.

(Btw, I edited my comment and added another open question 11 minutes ago. Not sure if you saw that :slightly_smiling_face:)

Marwes commented:
> If a reconnection attempt is in progress, should the current connection be considered dead? Or could it be that connection errors are transient? In the former case, commands should be delayed until the new connection is established, while in the latter case all commands should be forwarded to the old connection, with the possibility that they fail as well.

As I see it, as soon as a reconnection attempt is started there is no way to retrieve an old connection since the `ArcSwap` has been updated at that point.

dbrgn commented:
> As I see it, as soon as a reconnection attempt is started there is no way to retrieve an old connection since the `ArcSwap` has been updated at that point.

Ah, good point. In that case, as soon as a reconnection attempt is in progress, all requests should await the new connection.

dbrgn commented:
TIL that `ArcSwap` does not behave like a pointer and should not be cloned in our case :man_facepalming: :slightly_smiling_face: https://github.com/vorner/arc-swap/pull/24

I guess we should use `connection: Arc<ArcSwap<SharedRedisFuture<MultiplexedConnection>>>` for our connection.

I got a version with async reconnection and non-parallel reconnections to work, will have to clean up the code and push it later.

dbrgn commented:
Ok, I think now I have a version that does what it should! It's now ready to review.

Currently implemented behavior:

- When creating an instance of the `ConnectionManager`, an initial connection will be established and awaited. Connection errors will be returned directly.
- When a command sent to the server fails with an error that represents a "connection dropped" condition, that error will be passed on to the user, but it will trigger a reconnection in the background.
- The reconnect code will atomically swap the current (dead) connection with a future that will eventually resolve to a `MultiplexedConnection` or to a `RedisError`
- All commands that are issued after the reconnect process has been initiated, will have to await the connection future.
- If reconnecting fails, all pending commands will be failed as well. A new reconnection attempt will be triggered if the error is an I/O error.

This can be tested by decreasing the delay in `examples/async-connection-loss.rs` to a very low value like 1ms. (I removed most debug prints from the committed code, but left in a few lines.) Here's a test run (3 futures being resolved in parallel every 1ms):

Initially, the connection still works:

```
> PING
> PING
> PING
< Ok("PONG")
< Ok("PONG")
< Ok("PONG")
```

Then I restarted the Redis server. In the debug output, you can see the pointer to the connection future before and after being swapped.

If the log output contains "Swap happened", then a stale connection was replaced with a connection future. If the log output contains "Swap did not happen", then another task was faster in creating a connection future. In this case, no connection will be established.

```
> PING
> PING
> PING
Connection lost, reconnecting
Prepare reconnect
Swap happened
    Current: 0x558835c06c50
        New: 0x558835c077d0
Connection lost, reconnecting
Prepare reconnect
Swap did not happen
Connection lost, reconnecting
Prepare reconnect
Swap did not happen
< Err(Connection reset by peer (os error 104))
< Err(broken pipe)
< Err(broken pipe)
```

On the next command a few ms later, the server was not yet back up, so the reconnect future fails and is immediately replaced by a new connection future:

```
> PING
> PING
> PING
Reconnect failed, reconnecting
Prepare reconnect
Swap happened
    Current: 0x558835c077d0
        New: 0x558835c01a10
Reconnect failed, reconnecting
Prepare reconnect
Swap did not happen
Reconnect failed, reconnecting
Prepare reconnect
Swap did not happen
< Err(Reconnecting failed: Connection refused (os error 111))
< Err(Reconnecting failed: Connection refused (os error 111))
< Err(Reconnecting failed: Connection refused (os error 111))
```

Sometimes the reconnection future fails so fast that the next (concurrent) task already sees the new, failed future:

```
> PING
> PING
> PING
Reconnect failed, reconnecting
Prepare reconnect
Swap happened
    Current: 0x558835c02ab0
        New: 0x558835c02a10
Reconnect failed, reconnecting
Prepare reconnect
Swap happened
    Current: 0x558835c02a10
        New: 0x558835c07800
Reconnect failed, reconnecting
Prepare reconnect
Swap did not happen
< Err(Reconnecting failed: Connection refused (os error 111))
< Err(Reconnecting failed: Connection refused (os error 111))
< Err(Reconnecting failed: Connection refused (os error 111))
Reconnecting
```
After a few dozen milliseconds, the server is back up and the connection future resolves to a working `MultiplexedConnection`.

```
> PING
> PING
> PING
Reconnect failed, reconnecting
Prepare reconnect
Swap happened
Reconnecting
< Err(Reconnecting failed: Connection refused (os error 111))
< Ok("PONG")
< Ok("PONG")
```

Not yet implemented:

- [Connection caching](https://github.com/mitsuhiko/redis-rs/pull/278#issuecomment-581504576). Is it worth it?
- Throttling of reconnections. Maybe just delaying the reconnection attempt by ~50ms would help?

TODO:

- Remove debug print calls
- Squash commits

Note: The rustfmt failures happen because 1.41 introduced a slight change in formatting rules. Maybe the rustfmt check should be pinned to a specific Rust version. The code can then be reformatted when bumping the used Rust version.

dbrgn commented:
Rebased against master. I also added a commit that removes the debug prints (should be squashed anyways, but can be temporarily reverted for easier testing).

@Marwes is there anything else for me to do here?

Marwes commented on src/types.rs:
```diff
@@ -405,6 +405,33 @@ impl RedisError {
+            ErrorRepr::ExtensionError(ref code, ref detail) => {
+                ErrorRepr::ExtensionError(code.clone(), detail.clone())
+            }
+            ErrorRepr::IoError(ref e) => ErrorRepr::WithDescriptionAndDetail(
```
Can't this still be an `io::Error` (preserving the kind would be nice)

https://doc.rust-lang.org/std/io/struct.Error.html#method.new

```rust
io::Error::new(e.kind(), e.to_string())
```

dbrgn replied:
You are right, fixed!

Marwes commented on src/types.rs:
```diff
@@ -405,6 +405,33 @@ impl RedisError {
+    /// The `ioerror_description` parameter will be used as the static error
+    /// description in case an `IoError` is found.
+    #[cfg(feature = "connection-manager")] // Used to avoid "unused method" warning
+    pub(crate) fn clone_mostly(&self, ioerror_description: Option<&'static str>) -> Self {
```
This is always called with `Some` so the `Option` is redundant.

Marwes commented on examples/async-connection-loss.rs:
```diff
@@ -0,0 +1,87 @@
+    let mut interval = interval(Duration::from_millis(1000));
+    loop {
+        interval.tick().await;
+        println!("");
```
Clippy complains about this

Marwes commented on examples/async-connection-loss.rs:
```diff
@@ -0,0 +1,87 @@
+    let mut interval = interval(Duration::from_millis(100));
+    loop {
+        interval.tick().await;
+        println!("");
```
Ditto

dbrgn commented on src/client.rs:
```diff
@@ -89,6 +89,32 @@ impl Client {
+    /// the connection will be re-established automatically if possible. (TODO
+    /// dbrgn: Document what happens if reconnection attempts fail, and what happens
+    /// to commands that are invoked while a reconnection attempt is in
+    /// progress.)
```
Oh, I forgot a TODO in here. Will fix that.

Marwes commented:
Looks good! Thanks!

dbrgn commented:
On this occasion, thanks @Marwes for the fantastic mentoring and quick feedback you did here - I probably wouldn't have managed to get to this solution without you :slightly_smiling_face:

Terkwood commented:
Late to the party, but this looks like a great addition! Thank you!
